amethyst_renderer/src/lib.rs

```diff
@@ -0,0 +1,285 @@
+extern crate num_cpus;
+extern crate rayon;
+extern crate serde;
+#[macro_use]
+extern crate serde_derive;
+extern crate winit;
+
+#[cfg(all(feature = "d3d11", target_os = "windows"))]
+extern crate gfx_device_dx11;
+#[cfg(all(feature = "d3d11", target_os = "windows"))]
+extern crate gfx_window_dxgi;
+
+#[cfg(all(feature = "metal", target_os = "macos"))]
+extern crate gfx_device_metal;
+#[cfg(all(feature = "metal", target_os = "macos"))]
+extern crate gfx_window_metal;
+
+#[cfg(feature = "opengl")]
+extern crate gfx_device_gl;
+#[cfg(feature = "opengl")]
+extern crate gfx_window_glutin;
+#[cfg(feature = "opengl")]
+extern crate glutin;
+
+#[cfg(feature = "vulkan")]
+extern crate gfx_device_vulkan;
+#[cfg(feature = "vulkan")]
+extern crate gfx_window_vulkan;
+
+pub use cam::{Camera, Projection};
+pub use color::Rgba;
+pub use error::{Error, Result};
+pub use light::Light;
+pub use mesh::{Mesh, MeshBuilder};
+pub use mtl::{Material, MaterialBuilder};
+pub use pipe::{Pipeline, PipelineBuilder, Stage, Target};
+pub use scene::{LightIter, MeshIter, Scene};
+pub use tex::{Texture, TextureBuilder};
+pub use vertex::VertexFormat;
+
+use pipe::{ColorBuffer, DepthBuffer};
+use std::sync::Arc;
+use std::time::Duration;
+use types::{ColorFormat, DepthFormat, Encoder, Factory, Window};
+use winit::WindowBuilder;
+
+#[cfg(feature = "opengl")]
+use glutin::EventsLoop;
+#[cfg(not(feature = "opengl"))]
+use winit::EventsLoop;
+
+pub mod light;
+pub mod pass;
+pub mod prelude;
+pub mod pipe;
+pub mod vertex;
+
+mod cam;
+mod color;
+mod error;
+mod mesh;
+mod mtl;
+mod scene;
+mod tex;
+mod types;
+
+/// Generic renderer.
+pub struct Renderer {
+    device: types::Device,
+    encoders: Vec<Encoder>,
+    factory: Factory,
+    main_target: Arc<Target>,
+    pool: rayon::ThreadPool,
+    window: Window,
+}
+
+impl Renderer {
+    /// Creates a new renderer with default window settings.
+    pub fn new(el: &EventsLoop) -> Result<Renderer> {
+        let wb = WindowBuilder::new().with_title("Amethyst");
+        Renderer::from_winit_builder(wb, el)
+    }
+
+    /// Creates a new renderer with the given `winit::WindowBuilder`.
+    pub fn from_winit_builder(wb: WindowBuilder, el: &EventsLoop) -> Result<Renderer> {
+        let Backend(dev, mut fac, main, win) = init_backend(wb, el)?;
+
+        let num_cores = num_cpus::get();
+        let encoders = (0..num_cores)
+            .map(|_| fac.create_command_buffer().into())
+            .collect();
+
+        let cfg = rayon::Configuration::new().num_threads(num_cores);
+        let pool = rayon::ThreadPool::new(cfg)
+            .map_err(|e| Error::PoolCreation(e))?;
+
+        Ok(Renderer {
+            device: dev,
+            encoders: encoders,
+            factory: fac,
+            main_target: Arc::new(main),
+            pool: pool,
+            window: win,
+        })
+    }
+
+    /// Builds a new mesh from the given vertices.
+    pub fn create_mesh(&mut self, mb: MeshBuilder) -> Result<Mesh> {
```

Comment:
Ah oh I see, you're passing it in here. Maybe you could use `pub(crate)` for the builder then?